src/compiler/program.ts

```diff
@@ -171,7 +171,7 @@ namespace ts {
         const readFileCache = new Map<string, string | false>();
         const fileExistsCache = new Map<string, boolean>();
         const directoryExistsCache = new Map<string, boolean>();
-        const sourceFileCache = new Map<string, SourceFile>();
+        const sourceFileCache = new Map<string, ESMap<SourceFile["impliedNodeFormat"], SourceFile>>();

```

Comment:
Seems like you might end up with fewer map instances if impliedNodeFormat were the outer key.